d2compiler/compile.go

```diff
@@ -34,534 +32,409 @@ func Compile(path string, r io.RuneReader, opts *CompileOptions) (*d2graph.Graph
-			if err != nil {
-				c.errorf(mk.Range.Start, mk.Range.End, err.Error())
-				return
-			}
-			parent.AppendReferences(ida, d2graph.Reference{
-				Key: e.Src,
+	edge.Attributes.Label.MapKey = e.LastPrimaryKey()
+	for _, er := range e.References {
+		scopeObjIDA := d2ir.IDA(er.Context.ScopeMap)
+		scopeObj, _ := edge.Src.Graph.Root.HasChild(d2graphIDA(scopeObjIDA))
+		edge.References = append(edge.References, d2graph.EdgeReference{
+			Edge:            er.Context.Edge,
+			MapKey:          er.Context.Key,
+			MapKeyEdgeIndex: er.Context.EdgeIndex(),
+			Scope:           er.Context.Scope,
+			ScopeObj:        scopeObj,
+		})
+	}
+}
 
-				MapKey:          mk,
-				MapKeyEdgeIndex: i,
-				Scope:           m,
-			}, obj)
+func (c *compiler) compileEdgeField(edge *d2graph.Edge, f *d2ir.Field) {
+	keyword := strings.ToLower(f.Name)
+	_, isReserved := d2graph.SimpleReservedKeywords[keyword]
+	if isReserved {
+		c.compileReserved(edge.Attributes, f)
+		return
+	} else if f.Name == "style" {
+		if f.Map() == nil {
+			return
 		}
-		if e.Dst != nil {
-			ida := d2graph.Key(e.Dst)
-
-			parent, _, err := d2graph.ResolveUnderscoreKey(ida, obj)
-			if err != nil {
-				c.errorf(mk.Range.Start, mk.Range.End, err.Error())
-				return
-			}
-			parent.AppendReferences(ida, d2graph.Reference{
-				Key: e.Dst,
+		c.compileStyle(edge.Attributes, f.Map())
+		return
+	}
 
-				MapKey:          mk,
-				MapKeyEdgeIndex: i,
-				Scope:           m,
-			}, obj)
+	if f.Name == "source-arrowhead" || f.Name == "target-arrowhead" {
+		if f.Map() != nil {
+			c.compileArrowheads(edge, f)
 		}
 	}
 }
 
-func (c *compiler) compileFlatKey(k *d2ast.KeyPath) ([]string, string, bool) {
-	k2 := *k
-	var reserved string
-	for i, s := range k.Path {
-		keyword := strings.ToLower(s.Unbox().ScalarString())
-		_, isReserved := d2graph.ReservedKeywords[keyword]
-		_, isReservedHolder := d2graph.ReservedKeywordHolders[keyword]
-		if isReserved && !isReservedHolder {
-			reserved = keyword
-			k2.Path = k2.Path[:i]
-			break
-		}
+func (c *compiler) compileArrowheads(edge *d2graph.Edge, f *d2ir.Field) {
+	var attrs *d2graph.Attributes
+	if f.Name == "source-arrowhead" {
+		edge.SrcArrowhead = &d2graph.Attributes{}
+		attrs = edge.SrcArrowhead
+	} else {
+		edge.DstArrowhead = &d2graph.Attributes{}
+		attrs = edge.DstArrowhead
 	}
-	if len(k2.Path) < len(k.Path)-1 {
-		c.errorf(k.Range.Start, k.Range.End, "reserved key %q cannot have children", reserved)
-		return nil, "", false
+
+	if f.Primary() != nil {
+		c.compileLabel(attrs, f)
+	}
+
+	for _, f2 := range f.Map().Fields {
+		keyword := strings.ToLower(f2.Name)
+		_, isReserved := d2graph.SimpleReservedKeywords[keyword]
+		if isReserved {
+			c.compileReserved(attrs, f2)
+			continue
+		} else if f2.Name == "style" {
+			if f2.Map() == nil {
+				continue
+			}
+			c.compileStyle(attrs, f2.Map())
+			continue
+		} else {
+			c.errorf(f2.LastRef().AST(), `source-arrowhead/target-arrowhead map keys must be reserved keywords`)
```

Comment:
how can you be sure it's the last ref that's doing this?

Reply:
Any of the refs are doing it, I just use the last ref because it's most accessible.

Reply:
i don't understand what "any of the refs are doing it" means. u can have multiple refs of the source arrowhead and only one is doing the illegal thing right?

```
a -> b: { target-arrowhead.label: 1 }
(a -> b)[0].target-arrowhead.x: y
(a -> b)[0].target-arrowhead.shape: diamond
```

Reply:
f2.LastRef().AST() is the last ref of the offending key inside source-arrowhead/target-arrowhead.